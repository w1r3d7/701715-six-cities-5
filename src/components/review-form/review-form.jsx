import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import ReviewRating from '../review-rating/review-rating';

import {RATING_COUNTS, RATING_TITLES} from '../../constants/constants.js';
import {sendReview} from '../../store/data/api-actions';
import {getError, getReviewSendStatus} from '../../store/selectors';

const FormFieldName = {
  REVIEW: `review`,
  RATING: `rating`,
};

const ReviewForm = ({
  sendReviewAction,
  offerId,
  sendReviewError,
  reviewRequestStatus
}) => {
  const handleFormSubmit = (evt) => {
    evt.preventDefault();
    const formData = new FormData(evt.target);
    const areaText = formData.get(FormFieldName.REVIEW);
    const currentRating = Number(formData.get(FormFieldName.RATING));
    const review = {
      comment: areaText,
      rating: currentRating,
    };
    evt.target.disabled = true;
    sendReviewAction(offerId, review);
    evt.target.reset();
  };

  const errorBlock = sendReviewError ? <p style={{color: `red`}}>Error `{sendReviewError}`, please try later</p> : false;

  return (
    <form
      className="reviews__form form"
      action="#"
      method="post"
      onSubmit={handleFormSubmit}
    >
      <label className="reviews__label form__label" htmlFor="review">Your review</label>
      <div className="reviews__rating-form form__rating">
        {RATING_TITLES.map((title, index) => {
          const rating = RATING_COUNTS[index];
          return (
            <ReviewRating
              key={title}
              title={title}
              rating={rating}
              isDisabled={reviewRequestStatus}
            />
          );
        })}
      </div>
      <textarea className="reviews__textarea form__textarea"
        id="review"
        name="review"
        placeholder="Tell how was your stay, what you like and what can be improved"
        maxLength="300"
        minLength="50"
        disabled={reviewRequestStatus ? `disabled` : ``}
      />
      {errorBlock}
      <div className="reviews__button-wrapper">
        <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and describe your
            stay with at least <b className="reviews__text-amount">50 characters</b>.
        </p>
        <button
          className="reviews__submit form__submit button"
          type="submit"
          disabled={reviewRequestStatus ? `disabled` : ``}
        >Submit</button>
      </div>
    </form>
  );
};

ReviewForm.propTypes = {
  offerId: PropTypes.number.isRequired,
  sendReviewAction: PropTypes.func.isRequired,
  sendReviewError: PropTypes.string,
  reviewRequestStatus: PropTypes.bool.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  sendReviewAction: (id, review) => dispatch(sendReview(id, review))
});

const mapStateToProps = (state) => ({
  sendReviewError: getError(state),
  reviewRequestStatus: getReviewSendStatus(state),
});

export {ReviewForm};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewForm);
