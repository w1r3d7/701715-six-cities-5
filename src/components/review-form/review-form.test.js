import React from 'react';
import renderer from 'react-test-renderer';

import {ReviewForm} from './review-form';

const noop = () => {};

describe(`ReviewForm render correctly`, () => {
  it(`ReviewForm with error`, () => {
    const props = {
      offerId: 1,
      sendReviewAction: noop,
      reviewRequestStatus: false,
      sendReviewError: `404`,
    };

    const tree = renderer
      .create(
          <ReviewForm {...props} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`ReviewForm without error`, () => {
    const props = {
      offerId: 1,
      sendReviewAction: noop,
      reviewRequestStatus: false,
    };

    const tree = renderer
      .create(
          <ReviewForm {...props} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});


