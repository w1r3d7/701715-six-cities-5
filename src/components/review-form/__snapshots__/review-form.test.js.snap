// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm render correctly ReviewForm with error 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="5-stars"
      name="rating"
      required={true}
      type="radio"
      value="5"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="4-stars"
      name="rating"
      required={true}
      type="radio"
      value="4"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="3-stars"
      name="rating"
      required={true}
      type="radio"
      value="3"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="2-stars"
      name="rating"
      required={true}
      type="radio"
      value="2"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="1-stars"
      name="rating"
      required={true}
      type="radio"
      value="1"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled=""
    id="review"
    maxLength="300"
    minLength="50"
    name="review"
    placeholder="Tell how was your stay, what you like and what can be improved"
  />
  <p
    style={
      Object {
        "color": "red",
      }
    }
  >
    Error \`
    404
    \`, please try later
  </p>
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`ReviewForm render correctly ReviewForm without error 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="5-stars"
      name="rating"
      required={true}
      type="radio"
      value="5"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="4-stars"
      name="rating"
      required={true}
      type="radio"
      value="4"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="3-stars"
      name="rating"
      required={true}
      type="radio"
      value="3"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="2-stars"
      name="rating"
      required={true}
      type="radio"
      value="2"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      disabled=""
      id="1-stars"
      name="rating"
      required={true}
      type="radio"
      value="1"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled=""
    id="review"
    maxLength="300"
    minLength="50"
    name="review"
    placeholder="Tell how was your stay, what you like and what can be improved"
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
